src/core/execute.c

```diff
@@ -411,37 +468,109 @@ int exec_spawn(Unit *unit,
+                        fdset = fdset_new();
+                        if (!fdset)
+                                return log_oom();
+
+                        if (ftruncate(fileno(f), 0) < 0)
+                                return log_unit_error_errno(unit,
+                                                errno,
+                                                "Failed to truncate serialization stream: %m");
+
+                        r = exec_serialize_invocation(f,
+                                                      fdset,
+                                                      /* serialize_fd_index= */ false,
+                                                      context,
+                                                      command,
+                                                      params,
+                                                      runtime,
+                                                      cgroup_context);
+                        if (r < 0)
+                                return log_unit_error_errno(unit, r, "Failed to serialize parameters: %m");
+
+                        rewind(f);
+
+                        r = direct_spawn_fallback(unit, f, fdset, &pid);
+                        if (r < 0)
+                                return log_unit_error_errno(unit,
+                                                r,
+                                                "Failed to directly spawn executor: %m");
+                } else {
+                        char iov_buffer[DECIMAL_STR_MAX(int) + 1] = {};
+                        struct iovec iov = IOVEC_MAKE(iov_buffer, sizeof(iov_buffer) - 1);
+                        _cleanup_close_ int pidfd = -EBADF;
+
+                        ssize_t n = receive_one_fd_iov(unit->manager->executors_pool_socket[0],
+                                        &iov,
+                                        /* iovlen= */ 1,
+                                        /* flags= */ 0,
+                                        &pidfd);
+                        if (n < 0)
+                                return log_error_errno(n, "Failed to receive worker's PID: %m");
+
+                        /* In case we are running on a kernel without pidfd support */
+                        if (pidfd < 0) {
+                                r = parse_pid(iov_buffer, &pid);
+                                if (r < 0)
+                                        return log_error_errno(r,
+                                                        "Failed to parse worker's PID from '%s': %m",
+                                                        strna(iov_buffer));
+                        } else {
+                                r = pidfd_get_pid(pidfd, &pid);
+                                if (r < 0)
+                                        return log_error_errno(r,
+                                                        "Failed to get worker's PID from PID fd: %m");
+                        }
```

Comment:
please turn this into a function of its own: recv_worker_pid() or so, that ideally returns a PidRef even